### PR TITLE
[ webkitpy ] Generate results.html for other test types

### PR DESCRIPTION
**Pull request by Smackteo**
#### 418610153aa75831dddd771f46643cfe4d9f79a6
<pre>
[ webkitpy ] Generate results.html for other test types
<a href="https://rdar.apple.com/143930188">rdar://143930188</a>

Reviewed by Jonathan Bedard.

This adds a failsafe to the LayoutTests/fast/harness/results.html reference file.

* Tools/Scripts/webkitpy/layout_tests/controllers/manager.py:
(Manager._copy_results_html_file):

Canonical link: <a href="https://commits.webkit.org/289654@main">https://commits.webkit.org/289654@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/e035abda1065515d8c7927feed5fb343f0f9bffe

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/87597 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/7112 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/41978 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/92462 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/38341 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/89648 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/7493 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/15281 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/67645 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/25391 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/90599 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/5721 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/79267 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/48008 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/87094 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/5508 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/33661 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/37454 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/75910 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/34525 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/94348 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/14765 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/10882 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/76495 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/15019 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/75116 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/75717 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18627 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/20091 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/18520 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/7716 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/14781 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/14525 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/17969 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/16307 "Built successfully") | | |

### CODE REVIEW

File: Tools/Scripts/webkitpy/layout_tests/controllers/manager.py [hunk 1]
```diff
@@ -733,10 +733,15 @@ def _save_json_files(self, summarized_results, initial_results):
...
     def _copy_results_html_file(self, filename, destination_path):
         base_dir = self._port.path_from_webkit_base('LayoutTests', 'fast', 'harness')
+        backup_dir = os.path.abspath(os.path.join(os.path.dirname(__file__), '../../../../../', 'LayoutTests', 'fast', 'harness'))
         results_file = self._filesystem.join(base_dir, filename)
+        backup_file = self._filesystem.join(backup_dir, filename)
         # Note that the results.html template file won't exist when we're using a MockFileSystem during unit tests,
         # so make sure it exists before we try to copy it.
-        if self._filesystem.exists(results_file):
+        if not self._filesystem.exists(results_file):
+            if self._filesystem.exists(backup_file):
```
Comment (JonWBedard) [inline Tools/Scripts/webkitpy/layout_tests/controllers/manager.py]: Rather than this nested check, we should do:
```
if not self._filesystem.exists(results_file):
    backup_dir = os.path.abspath(os.path.join(os.path.dirname(__file__), '../../../../../', 'LayoutTests', 'fast', 'harness'))
    results_file = self._filesystem.join(backup_dir, filename)
if self._filesystem.exists(results_file):
    self._filesystem.copyfile(results_file, destination_path)
```

I think that will be easier to understand, rather than this nested if statement thing

Comment (Smackteo) [reply Tools/Scripts/webkitpy/layout_tests/controllers/manager.py]: > Rather than this nested check, we should do:
> 
> ```
> if not self._filesystem.exists(results_file):
>     backup_dir = os.path.abspath(os.path.join(os.path.dirname(__file__), '../../../../../', 'LayoutTests', 'fast', 'harness'))
>     results_file = self._filesystem.join(backup_dir, filename)
> if self._filesystem.exists(results_file):
>     self._filesystem.copyfile(results_file, destination_path)
> ```
> 
> I think that will be easier to understand, rather than this nested if statement thing

I agree that is easier to read. Making that change.

**Review by JonWBedard — APPROVED** [review]

**Comment by gsnedders** [comment]
Too late, but:

This is needed because `Port.path_from_webkit_base` does a lot of silliness if the first argument is `LayoutTests`, right? i.e.:

https://github.com/WebKit/WebKit/blob/909da23bf3679f00fe7f8bc2e3a5b60dfc658f80/Tools/Scripts/webkitpy/port/base.py#L582-L583

 Would just changing the definition of `base_dir` from:
 
 ```patch
-         base_dir = self._port.path_from_webkit_base('LayoutTests', 'fast', 'harness')
+         base_dir = self._filesystem.join(self._port.webkit_base(), 'LayoutTests', 'fast', 'harness')
```
 
Not solve this without having to hardcode finding the path of the WebKit root, again?

(Also, still for now we shouldn't be using `os.path.join` and the like — we should be using `self._filesystem.join` and other things defined on our [FileSystem class](https://github.com/WebKit/WebKit/blob/909da23bf3679f00fe7f8bc2e3a5b60dfc658f80/Tools/Scripts/webkitpy/common/system/filesystem.py), which admittedly mostly matters for tests. It also would've been nice to have a test for this, even if was only testing `_copy_results_html_file` in isolation, because this is the exact sort of edge-case behaviour we're likely to break.)
